### PR TITLE
Skip empty parameters in gradient reduction

### PR DESCRIPTION
**Pull request by tohtana**
#7736 fixed an issue with OnebitLamb NaN propagation. With the fix, the optimizer correctly filters out empty parameters, but DeepSpeed engine's gradient allreduce operation (which runs separately from the optimizer) still includes empty parameters' gradients.

This PR addresses the issue by skipping empty parameters (numel=0) in `_get_gradients_for_reduction()`.

### CODE REVIEW

File: deepspeed/runtime/engine.py [hunk 1]
```diff
@@ -2940,6 +2940,11 @@ def _get_gradients_for_reduction(self):
             if not param.requires_grad:
                 continue
 
+            # Skip empty parameters (numel=0) as they contribute nothing to gradient reduction
+            # and cause issues with flatten/unflatten operations
+            if param.numel() == 0:
```
Comment (PKUWZP) [inline deepspeed/runtime/engine.py:2945]: @tohtana Very clean fix! The only minor comment is that maybe we can add an explicit test for gradient reduction? It's optional though.

Comment (tohtana) [reply deepspeed/runtime/engine.py:2945]: Thank you for the feedback. Can you elaborate what test you are suggesting? Run only gradient reduction?

Comment (tohtana) [reply deepspeed/runtime/engine.py:2945]: Let me merge this PR for now. Please open a new issue or PR for the additional test.
